### ISSUE TITLE
Crash on launch since v2.0.0

### ISSUE THREAD

**Issue by rwg** [freeotp/freeotp-android#332]
I'm currently running FreeOTP v2.0.1 from the Google Play Store on a Google Pixel 3a. After upgrading from v1.x to v2.0.0, FreeOTP required that I set a backup password, and I was able to create a backup file. On every launch of FreeOTP since that first successful v2.0.0 launch when I created the backup file, including launches with v2.0.1, FreeOTP crashes immediately upon launch. Versions before v2.0.0 worked without issue.

This is the exception causing the crash, obtained with `adb logcat`:

```
FATAL EXCEPTION: main
Process: org.fedorahosted.freeotp, PID: 1592
java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String org.fedorahosted.freeotp.Token.getImage()' on a null object reference
	at org.fedorahosted.freeotp.TokenIcon.getImage(TokenIcon.java:81)
	at org.fedorahosted.freeotp.TokenIcon.<init>(TokenIcon.java:37)
	at org.fedorahosted.freeotp.main.Adapter.onBindViewHolder(Adapter.java:169)
	at org.fedorahosted.freeotp.main.Adapter.onBindViewHolder(Adapter.java:64)
	at androidx.recyclerview.widget.RecyclerView$Adapter.onBindViewHolder(RecyclerView.java:7065)
	at androidx.recyclerview.widget.RecyclerView$Adapter.bindViewHolder(RecyclerView.java:7107)
	at androidx.recyclerview.widget.RecyclerView$Recycler.tryBindViewHolderByDeadline(RecyclerView.java:6012)
	at androidx.recyclerview.widget.RecyclerView$Recycler.tryGetViewHolderForPositionByDeadline(RecyclerView.java:6279)
	at androidx.recyclerview.widget.RecyclerView$Recycler.getViewForPosition(RecyclerView.java:6118)
	at androidx.recyclerview.widget.RecyclerView$Recycler.getViewForPosition(RecyclerView.java:6114)
	at androidx.recyclerview.widget.LinearLayoutManager$LayoutState.next(LinearLayoutManager.java:2303)
	at androidx.recyclerview.widget.GridLayoutManager.layoutChunk(GridLayoutManager.java:561)
	at androidx.recyclerview.widget.LinearLayoutManager.fill(LinearLayoutManager.java:1587)
	at androidx.recyclerview.widget.LinearLayoutManager.onLayoutChildren(LinearLayoutManager.java:665)
	at androidx.recyclerview.widget.GridLayoutManager.onLayoutChildren(GridLayoutManager.java:170)
	at androidx.recyclerview.widget.RecyclerView.dispatchLayoutStep2(RecyclerView.java:4134)
	at androidx.recyclerview.widget.RecyclerView.dispatchLayout(RecyclerView.java:3851)
	at androidx.recyclerview.widget.RecyclerView.onLayout(RecyclerView.java:4404)
	at android.view.View.layout(View.java:23203)
	at android.view.ViewGroup.layout(ViewGroup.java:6412)
	at com.google.android.material.appbar.HeaderScrollingViewBehavior.layoutChild(HeaderScrollingViewBehavior.java:148)
	at com.google.android.material.appbar.ViewOffsetBehavior.onLayoutChild(ViewOffsetBehavior.java:43)
	at com.google.android.material.appbar.AppBarLayout$ScrollingViewBehavior.onLayoutChild(AppBarLayout.java:2122)
	at androidx.coordinatorlayout.widget.CoordinatorLayout.onLayout(CoordinatorLayout.java:918)
	at android.view.View.layout(View.java:23203)
	at android.view.ViewGroup.layout(ViewGroup.java:6412)
	at android.widget.FrameLayout.layoutChildren(FrameLayout.java:332)
	at android.widget.FrameLayout.onLayout(FrameLayout.java:270)
	at android.view.View.layout(View.java:23203)
	at android.view.ViewGroup.layout(ViewGroup.java:6412)
	at android.widget.LinearLayout.setChildFrame(LinearLayout.java:1829)
	at android.widget.LinearLayout.layoutVertical(LinearLayout.java:1673)
	at android.widget.LinearLayout.onLayout(LinearLayout.java:1582)
	at android.view.View.layout(View.java:23203)
	at android.view.ViewGroup.layout(ViewGroup.java:6412)
	at android.widget.FrameLayout.layoutChildren(FrameLayout.java:332)
	at android.widget.FrameLayout.onLayout(FrameLayout.java:270)
	at android.view.View.layout(View.java:23203)
	at android.view.ViewGroup.layout(ViewGroup.java:6412)
	at android.widget.LinearLayout.setChildFrame(LinearLayout.java:1829)
	at android.widget.LinearLayout.layoutVertical(LinearLayout.java:1673)
	at android.widget.LinearLayout.onLayout(LinearLayout.java:1582)
	at android.view.View.layout(View.java:23203)
	at android.view.ViewGroup.layout(ViewGroup.java:6412)
	at android.widget.FrameLayout.layoutChildren(FrameLayout.java:332)
	at android.widget.FrameLayout.onLayout(FrameLayout.java:270)
	at com.android.internal.policy.DecorView.onLayout(DecorView.java:797)
	at android.view.View.layout(View.java:23203)
	at android.view.ViewGroup.layout(ViewGroup.java:6412)
	at android.view.ViewRootImpl.performLayout(ViewRootImpl.java:3749)
	at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:3205)
	at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:2179)
	at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:8787)
	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1037)
	at android.view.Choreographer.doCallbacks(Choreographer.java:845)
	at android.view.Choreographer.doFrame(Choreographer.java:780)
	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:1022)
	at android.os.Handler.handleCallback(Handler.java:938)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loopOnce(Looper.java:201)
	at android.os.Looper.loop(Looper.java:288)
	at android.app.ActivityThread.main(ActivityThread.java:7870)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)
```

**Comment by justin-stephenson** [comment]
Hi,

Somehow the token object retrieved in [onBindViewHolder](https://github.com/freeotp/freeotp-android/blob/master/mobile/src/main/java/org/fedorahosted/freeotp/main/Adapter.java#L166) is null. The token metadata is retrieved from the shared preferences file named `tokenStore` in the FreeOTP app private data directory. The contents of this file look should similar to below, I believe only rooted devices can view this file however.

~~~
<?xml version='1.0' encoding='utf-8' standalone='yes' ?>
<map>
    <string name="2a4bd3f9-922d-42ea-a359-c160ca501558">{&quot;algo&quot;:&quot;SHA256&quot;,&quot;counter&quot;:0,&quot;digits&quot;:6,&quot;issuerExt&quot;:&quot;test&quot;,&quot;label&quot;:&quot;test&quot;,&quot;lock&quot;:false,&quot;period&quot;:30,&quot;type&quot;:&quot;HOTP&quot;}</string>
    <string name="tokenOrder">[&quot;2a4bd3f9-922d-42ea-a359-c160ca501558&quot;]</string>
</map>
~~~

If you can copy the `tokenBackup.xml` shared preferences file from the FreeOTP private app data directory elsewhere, then you can try uninstalling FreeOTP and copy the file back to the phone, then inside FreeOTP you can select the 'Restore' option in the top right and browse to this backup file.

**Comment by rwg** [comment]
Hi @justin-stephenson,

I ended up buying a new phone (not because of this issue), and I was able to import the FreeOTP backup I'd made on the old phone into FreeOTP on the new phone.

My old phone isn't rooted, so I'm not able to get at the `tokenStore` file you mentioned with `adb` and see if that file has any clues about what's causing the crash. I'm going to go on and close this issue since there's not much I can do to troubleshoot this on my side.
